### PR TITLE
Disable superfluous internal checks systematically. 

### PR DESCRIPTION
**Pull request by HechtiDerLachs**
As noted by @simonbrandhorst some code for resolutions of singularities of elliptic fibrations was horribly slow. Today, I have been going through the code and systematically disabled all superfluous internal checks that were triggered by our test file. 

To this end, I introduced a flag called `CHECK_ERROR`. I will point out the description in the source code below. This is probably not in its optimal form, but it was the best I could do for the moment. And it has worked quite well for me so far. 

@thofma @fingolfin : Could you have a look at this design and let me know what you think? Maybe one should solve this differently, but after what I saw today, I think we definitely need this functionality in one way or the other. 

@thofma : I also changed some parts in your code on the maps from polynomial rings. Have a look please, whether you approve.

### CODE REVIEW

File: src/AlgebraicGeometry/Schemes/main.jl
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/main.jl]: This is the flag I was talking about.

Comment (simonbrandhorst) [reply src/AlgebraicGeometry/Schemes/main.jl]: Surely there is a better place to put this?
My impression is that this makes sense for prohibitively expensive checks in the schemes world. 
But not as a general pattern in Oscar.

File: src/AlgebraicGeometry/Schemes/main.jl
Comment (simonbrandhorst) [inline src/AlgebraicGeometry/Schemes/main.jl]:
```suggestion
# A function using the optional `check::Bool=true` argument can have the line 
```

**Comment by simonbrandhorst** [comment]
Disabling the checks is fine.  Testbots fail.

File: src/AlgebraicGeometry/Schemes/main.jl
Comment (HechtiDerLachs) [reply src/AlgebraicGeometry/Schemes/main.jl]: I thought so, too. But as I progressed, I saw this spreading more and more into the algebra section. So I do think this should become a more general thing, unfortunately. I just didn't know where to put it and wanted to get some opinions first.

**Comment by thofma** [comment]
I understand what this is trying to do, but I must say, that all these `CHECK_ERROR && error("check was enabled")` look awful.

It should be `const CHECK_ERROR`.

**Comment by HechtiDerLachs** [comment]
I know, it looks awful. I just needed something quick and dirty to work with today. We should change it, but I wanted to hear some opinions on how.

File: experimental/Schemes/Types.jl
Comment (afkafkafk13) [inline experimental/Schemes/Types.jl]: I do not understand, why line 150 was introduced. Is this a leftover from searching all unnecessary tests?

Comment (simonbrandhorst) [reply experimental/Schemes/Types.jl]: It is to facilitate catching unnecessary tests in the future.

Comment (afkafkafk13) [reply experimental/Schemes/Types.jl]: ok. got it -- when I started reading the changes (and was interrupted many times), the discussion here had not yet started.

File: src/AlgebraicGeometry/Schemes/main.jl
Comment (simonbrandhorst) [inline src/AlgebraicGeometry/Schemes/main.jl]:
```suggestion
const CHECK_ERROR = false
```

Comment (HechtiDerLachs) [reply src/AlgebraicGeometry/Schemes/main.jl]: I think we can change this once people have agreed on how they want to address the "awfulness" of the `check && CHECK_ERROR && error(...)` line in general, whether all of this should be moved to another place, or whether we can leave it as is for now. No need to re-run all the tests in the meantime, unless they do something useful like heating a building which actually needs heating.

**Comment by fieker** [comment]
I don't think this is a good idea at all. It is going to bite you terribly if seriously implemented "everywhere": you're going to set this to false globally and then goging to run, via illegal input to say number fields, into interesting errors everywhere.
We've tried that before...

**Comment by fieker** [comment]
Talking to Mathias: I misunderstood. This idea is only to notify when input validation is happening with the aim to not use input validation internally

**Comment by HechtiDerLachs** [comment]
I have added a new macro in `src/Assertions.jl` similar to what I found in Hecke. You can also use it without any arguments which might seem a bit odd at first. Then it only makes sure the error message is thrown when snooping for unnecessary internal checks. Let me know what you think.

File: src/Oscar.jl
Comment (HechtiDerLachs) [inline src/Oscar.jl]: I was not sure where to put this. Let me know if there is a more appropriate place.

Comment (fingolfin) [reply src/Oscar.jl]: place is fine, but the filename should be lowercase

File: src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl
Comment (fingolfin) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl]: Huh, so that's weird: why does `check` control whether we test for empty `V` here, and lead to a different (??) return value? That looks as if it might be either an optimization (but then why is it restricted to `check == true`), or else a semantical change (but that would seem very undesirable). What am I missing?

File: experimental/Schemes/CoveredScheme.jl
Comment (fingolfin) [inline experimental/Schemes/CoveredScheme.jl]: This function just passes `check` on to other functions, so I don't think it needs `@check`, provided those functions use it?

File: experimental/Schemes/WeilDivisor.jl
Comment (fingolfin) [inline experimental/Schemes/WeilDivisor.jl]: I rather dislike these stand-alone `@check` invocations. I find them very hard to model right mentally. I would find it much more natural if this code read as
```suggestion
    @check do
```

File: experimental/Schemes/WeilDivisor.jl
Comment (fingolfin) [inline experimental/Schemes/WeilDivisor.jl]: the `check` argument is completely ignored by this method, why does it have it then? (also, if you add it here, shouldn't it also be added to the docstring above?

File: experimental/Schemes/WeilDivisor.jl
Comment (fingolfin) [inline experimental/Schemes/WeilDivisor.jl]: again: `check` unused and not mentioned in docstring

File: src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Constructors.jl
Comment (fingolfin) [inline src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Constructors.jl]: It seems to me this is an incorrect use of `check`, or at least different from other uses of `check`: instead of checking that the input is valid, it extends the set of valid inputs by reducing the input first. This could be a source of bad bugs. So either `check` should check for whether `X` is reduced, and if not, raise an error; or else it should be renamed to something else.

File: src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Constructors.jl
Comment (fingolfin) [inline src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Constructors.jl]: Here again `check` does not validate the input but rather modifies behavior: not good

File: src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Types.jl
Comment (fingolfin) [inline src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Types.jl]: Another opportunity to avoid the "stand-alone" `@check`:
```suggestion
    @check is_geometrically_reduced(X) "algebraic sets must be geometrically reduced"
```
(and then adjust the following code, too)

File: src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Attributes.jl
Comment (fingolfin) [inline src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Attributes.jl]: Again: if `check` is true then arguments should be validated, not "made to fit"

File: src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Attributes.jl
Comment (fingolfin) [inline src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Attributes.jl]: Is `@check true` any different from `@check`?

This is one of several methods which have unused `check` arguments. Why is that so? Perhaps this is meant to enable potential future extensions of this methods which use `check` after all?

File: src/Assertions.jl
Comment (lgoettgens) [inline src/Assertions.jl]: Is the compiler smart enough to optimize lines like this away, if the flag is `false`?
If not, would it be beneficial for performance to move the check outside, i.e. define different macros depending on the flag? Even if not necessary, I could imagine this being more readable.

File: src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl
Comment (afkafkafk13) [reply src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl]: You are missing that the emptyness test and the appropriate return value are part of the mathematics and the algorithm itself.
Testing isempty for non-empty V is not cheap (Groebner basis!), unless the value has already been cached. But depending on the setting, in which one calls the function, it is often inevitable to do this test. There are however numerous settings in which it is known beforehand from the construction that V is non-empty -- without having called isempty on V before. So check=false is an optimization here, but not in the direction that you were expecting.

File: src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Attributes.jl
Comment (afkafkafk13) [reply src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Attributes.jl]: You are right, mathematically the wording for 'check' is misleading here. It should be something like 'bypass_tayloring' (corresponding to the value of !check). Again this opens the possibility to bypass a Groebner basis computation, if the input is already known to satisfy the condition.

File: src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Constructors.jl
Comment (afkafkafk13) [reply src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Constructors.jl]: You are right, mathematically the wording for 'check' is misleading. It should be something like 'bypass_tayloring' (corresponding to the value of !check). This opens the possibility to bypass the rather expensive radical computation, if the input is already known to satisfy the condition.

File: src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Constructors.jl
Comment (afkafkafk13) [reply src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Constructors.jl]: You are right, mathematically the wording for 'check' is misleading here. It should be something like 'bypass_tayloring' (corresponding to the value of !check).

**Comment by thofma** [comment]
When I suggested the name `@check`, I thought we would use it as
```
@check blabla [optional error message]
```
I agree with @fingolfin that the single `@check` at the top of the function looks odd.

File: src/Oscar.jl
Comment (HechtiDerLachs) [reply src/Oscar.jl]: It was uppercase in Hecke. But I can change it.

File: src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl
Comment (HechtiDerLachs) [reply src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl]: For the final outcome does not make a difference: `isempty` being `true` implies `iszero(one(OO(V)) == true`. So it can not return anything but zero, regardless of the check being enabled or not. But yes: In general, we want to avoid the Groebner basis computation.

File: experimental/Schemes/CoveredScheme.jl
Comment (HechtiDerLachs) [reply experimental/Schemes/CoveredScheme.jl]: It doesn't need it, correct. However, this is a way to catch a faulty passing on of `check` at an earlier stage, so I included it.

File: experimental/Schemes/WeilDivisor.jl
Comment (HechtiDerLachs) [reply experimental/Schemes/WeilDivisor.jl]: I agree that the standalone check is rather awkward, especially in conjunction with another `if check` - block. I was not aware of the possibility to do a `do` block. It looks way more reasonable.

File: src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Constructors.jl
Comment (HechtiDerLachs) [reply src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Constructors.jl]: This is rather recent code of @simonbrandhorst . We should consult with him.

File: src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Attributes.jl
Comment (HechtiDerLachs) [reply src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Attributes.jl]: > Is `@check true` any different from `@check`?

No. It is just a deprecated way in which I was writing it at first. 

I can't say anymore why there are so many methods which apparently don't use `check`. But sometimes it's also just easier to add that optional argument to all methods of a given function, even though only some use it. For instance: The `hom(A, B, img_gens)` methods to create maps of rings has a `check` argument. We want to write generic code which accepts all combinations of types `MPolyRing`, `MPolyQuoRing`, `MPolyLocRing`, `MPolyQuoLocRing` for `A` and `B`. Whenever there is a modulus involved in the domain `A`, there is some well-definedness-check which is eventually to be skipped. So we write the code as `hom(A, B, img_gens, check=false)` for internal calls with sane input. But then also the method with signature `A::MPolyRing` needs to have a dummy optional argument `check::Bool`, even though there's nothing to do there. 

Does that make sense?

File: src/Assertions.jl
Comment (HechtiDerLachs) [reply src/Assertions.jl]: I'm not sure, I understand. Maybe we can talk on Slack/Gather? I'd be interested to hear more about this.

**Comment by HechtiDerLachs** [comment]
The signature you describe exists and is used. I will try to replace the standalone cases.

File: src/Assertions.jl
Comment (thofma) [reply src/Assertions.jl]: It does not really matter for performance, so I would not bother trying to change it. It will also be more tedious to make it a runtime option if the the check is moved outside.

**Comment by HechtiDerLachs** [comment]
I removed the standalone statements of `@check` up to the controversial files introduced by @simonbrandhorst .

File: src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl
Comment (fingolfin) [reply src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl]: OK, I was indeed missing this! Which to me suggest that a (brief!) comment explaining this would be helpful :-)

File: experimental/Schemes/CoveredScheme.jl
Comment (fingolfin) [reply experimental/Schemes/CoveredScheme.jl]: OK... Perhaps then we can come up with a better name for the `@check` macro, at least the zero argument version? As it is, the name to me suggests that it checks something, and for the other variants of the `@check` macro, that's true. Moreover, the difference to `@assert` or `@req` etc. is that it also takes the value of the `check` variable into account.

And then we tack on this extra check which makes it complain about `check = true` -- as I understand it, this is something purely for developers, meant to help them catch places that "forgot" to disable `check`. 

The zero-argument version of `@check` then does nothing in "normal" mode. Only in the "developer looking for forgotten `check=false` mode does it do something. So perhaps `@debug_check` would make more sense for the standalone version?

All that said, I am not going to block this PR over this nitpicking. I am just worried that code using `@check` gets harder to reason about.

So what I *really* want then is at least a good docstring for the macro that explain all this.

File: experimental/Schemes/WeilDivisor.jl
Comment (fingolfin) [reply experimental/Schemes/WeilDivisor.jl]: Correction: a `do` block won't work, what I should have suggested is a `begin`-block. I've already communicate this to @HechtiDerLachs via Slack, but I wanted to clarify it here, too, in case I am confusing any other readers. Sorry for that :-(

File: src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Attributes.jl
Comment (fingolfin) [reply src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Attributes.jl]: Yeah, that makes sense. Thanks

File: src/Oscar.jl
Comment (fingolfin) [reply src/Oscar.jl]: yeah Hecke, AA, Nemo (still) do many things a bit different, and also Oscar is not yet consistent, but we kinda agree that filenames should use `snake_case` and not `kebab-case` or `CamelCase`. But of course we have dozens and dozens of files with existing names; there is no strong reason to rename them all; and if inside a directory *all* files already use `CamelCase`, I am fine with new files also using that. But for the files directly inside `src`, we are using `snake_case`, so this file should as well (of course `src/Oscar.jl` is the sole exception, but we don't have a choice there).

File: src/Rings/MPolyMap/MPolyQuo.jl
Comment (fingolfin) [inline src/Rings/MPolyMap/MPolyQuo.jl:98]:
```suggestion
    _check_homo(S, imgs)
    true
```
But actually, I would suggest a different approach: drop these `true` again (readers may wonder why they are they, plus they are very easy to forget, as this example shows), and instead improve the `check` macro to better deal with these `begin` blocks.

Two options come to mind:
1. we could simply accept the return value `nothing` just like we accept `true`; the easiest way would be to replace `$(esc(cond)) || error($(esc(msg)))` by `$(esc(cond)) != false || error($(esc(msg)))`. This way, only something that really returns `false` constitutes a failed check.
2. we could literally check for `begin` expressions and treat those differently than other expressions... not sure that's a good idea, though...

File: src/Assertions.jl
Comment (fingolfin) [inline src/Assertions.jl]: But `@assert` can be turned off; yet in the next macro, you use
```
$(esc(cond)) || error($(esc(msg)))
```
which cannot be turned off like other assertions. I don't think users of the `@check` macro would expect that adding a message turns it from an assertion into a non-assertion...

So either both should use `@assert` (the second macro then in the two argument form `@assert cond msg`, or both should not use `@assert` (and thus *always* evaluate `cond`. I think the latter is more appropriate here, isn't it?

**Comment by fingolfin** [comment]
BTW, lest all my complaining and whining makes it sound as if I am opposed, let me emphasise that I think this is overall a good idea and I am grateful @HechtiDerLachs works on it. Thank you!

File: src/Oscar.jl
Comment (thofma) [reply src/Oscar.jl]: Even if the filename is the name of a type? I somehow missed that discussion.

**Comment by HechtiDerLachs** [comment]
All files by @simonbrandhorst have been reverted. He will take care of them later. From my side this should be ready to go.

File: experimental/Schemes/Types.jl
Comment (afkafkafk13) [inline experimental/Schemes/Types.jl:365]: This line is unreached, as the whole function is deactivated in line 352. Is this intentional and do you intentionally do changes in this unreached part?

File: src/AlgebraicGeometry/Schemes/Covering/Morphisms/Types.jl
Comment (afkafkafk13) [inline src/AlgebraicGeometry/Schemes/Covering/Morphisms/Types.jl:40]: Is this orphaned '@check' on purpose or was it forgotten in the last pass through the file?

File: src/AlgebraicGeometry/Schemes/ProjectiveAlgebraicSet/Objects/Constructors.jl
Comment (afkafkafk13) [inline src/AlgebraicGeometry/Schemes/ProjectiveAlgebraicSet/Objects/Constructors.jl:28]: again an orphaned '@check'... and one more further down in the file.

File: src/AlgebraicGeometry/Schemes/ProjectiveAlgebraicSet/Objects/Types.jl
Comment (afkafkafk13) [inline src/AlgebraicGeometry/Schemes/ProjectiveAlgebraicSet/Objects/Types.jl:4]: orphaned '@check'...

File: src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Types.jl
Comment (afkafkafk13) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Types.jl:24]: typo: lie instead of lay

**Review by afkafkafk13 — COMMENTED** [review]
Looks good to me. Mind the minor comments..

Thanks for this huge amount of work!

**Review by simonbrandhorst — APPROVED** [review]

File: src/AlgebraicGeometry/Schemes/Covering/Morphisms/Types.jl
Comment (simonbrandhorst) [reply src/AlgebraicGeometry/Schemes/Covering/Morphisms/Types.jl:40]: the check parameter is ignored too, but it is there for consistency.

**Comment by simonbrandhorst** [comment]
Since all the comments have been addressed, I will merge this in 24 hours unless someone speaks up.

**Pull request merged**
